### PR TITLE
Refactoring LAO Channel

### PR DESCRIPTION
**Pull request by MaximeZmt**
# Refactoring Lao Channel and the Registry

In this PR there are the following changes:
- The addition of a registry in the LAO channel
- Changing the registry to have as argument the socket of the sender, like that it can be reused to send something back.

### CODE REVIEW

File: be1-go/channel/consensus/mod.go [hunk 1]
```diff
@@ -438,7 +438,8 @@ func (c *Channel) createConsensusInstance(instanceID string) *ConsensusInstance
 }
 
 // processConsensusElect processes an elect action.
-func (c *Channel) processConsensusElect(message message.Message, msgData interface{}) error {
+
```
Comment (Robingoumaz) [inline be1-go/channel/consensus/mod.go:441]: You left a empty space

File: be1-go/channel/lao/mod.go [hunk 2]
```diff
@@ -213,16 +240,15 @@ func (c *Channel) verifyMessage(msg message.Message) error {
...
+// createGeneralChirpingChannel creates a new general chirping channel and returns it
 func createGeneralChirpingChannel(laoID string, hub channel.HubFunctionalities, socket socket.Socket) *generalChirping.Channel {
 	generalChannelPath := laoID + social + chirps
 	generalChirpingChannel := generalChirping.NewChannel(generalChannelPath, hub, be1_go.Logger)
 	hub.NotifyNewChannel(generalChannelPath, generalChirpingChannel, socket)
 
 	log.Info().Msgf("storing new channel '%s' ", generalChannelPath)
 
-	return generalChirpingChannel
+	return generalChirpingChannel //.(*generalChirping.Channel)
```
Comment (Robingoumaz) [inline be1-go/channel/lao/mod.go:251]: Is the comment supposed to be left here ?

**Review by Robingoumaz — COMMENTED** [review]
Good work just a few comments/questions

**Review by Warhedgehog — DISMISSED** [review]
LGTM !

**Review by Robingoumaz — DISMISSED** [review]

File: be1-go/channel/lao/mod.go [hunk 3]
```diff
@@ -425,111 +410,31 @@ func compareLaoUpdateAndState(update messagedata.LaoUpdate, state messagedata.La
...
+	err = schnorr.VerifyWithChecks(crypto.Suite, []byte(msg.Sender), []byte(witnessData.MessageID), []byte(witnessData.Signature))
 	if err != nil {
-		return answer.NewErrorf(-4, keyUnmarshalError, err)
-	}
-
-	if !c.organizerPubKey.Equal(senderPoint) {
-		return answer.NewErrorf(-5, "sender's public key %q does not match the organizer's", msg.Sender)
-	}
-
-	switch action {
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:490]: bye bye, ugly code 👋😄

File: be1-go/channel/lao/mod.go [hunk 4]
```diff
@@ -107,6 +110,28 @@ func NewChannel(channelID string, hub channel.HubFunctionalities, msg message.Me
...
+
+// NewLAORegistry creates a new registry for the LAO channel
+func (c *Channel) NewLAORegistry() registry.MessageRegistry {
+	registry := registry.NewMessageRegistry()
+
+	registry.Register(messagedata.LaoState{}, c.processLaoState)
+	registry.Register(messagedata.LaoUpdate{}, c.processEmptyFun)
+	registry.Register(messagedata.MeetingCreate{}, c.processEmptyFun)
+	registry.Register(messagedata.MeetingState{}, c.processEmptyFun)
+	registry.Register(messagedata.MessageWitness{}, c.processMessageObject)
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:127]: would make more sense, no ? 
```suggestion
	registry.Register(messagedata.MessageWitness{}, c.processMessageWitness)
```

File: be1-go/channel/lao/mod_test.go [hunk 5]
```diff
@@ -635,7 +635,7 @@ func (h *fakeHub) SendAndHandleMessage(msg method.Broadcast) error {
 	return nil
 }
 
-// fakeSocket is a fake implementation of a socket
+// fakeSocket is a fake implementation of a sockSocket
```
Comment (nkcr) [inline be1-go/channel/lao/mod_test.go:638]: really ? It is indeed a fake implementation of `Socket`. What is `sockSocket` ?

File: be1-go/channel/lao/mod.go [hunk 6]
```diff
@@ -675,10 +579,15 @@ func (c *Channel) processRollCallOpen(msg message.Message, action string) error
...
+func (c *Channel) processRollCallClose(msg message.Message, msgData interface{}, senderSocket socket.Socket) error {
+
+	data, ok := msgData.(*messagedata.RollCallClose)
+	if !ok {
+		return xerrors.Errorf("message %v isn't a rollcall#close message", msgData)
+	}
 
 	// check that data is correct
-	err := c.verifyMessageRollCallClose(msg)
+	err := c.verifyMessageRollCallClose(*data)
```
Comment (nkcr) [inline be1-go/channel/lao/mod.go:590]: you can update the verify functions to take pointers. At this point it doesn't make sense to create a copy.

File: be1-go/channel/mod.go [hunk 7]
```diff
@@ -69,7 +69,6 @@ func (s *Sockets) SendToAll(buf []byte) {
 func (s *Sockets) Upsert(socket socket.Socket) {
 	s.Lock()
 	defer s.Unlock()
-
```
Comment (nkcr) [inline be1-go/channel/mod.go:72]: indeed this return  was ok :)

**Review by nkcr — COMMENTED** [review]
Pretty neat 👌 
Added a couple of minor comments.

File: be1-go/channel/lao/mod.go (hunk 4 above)
Comment (MaximeZmt) [reply be1-go/channel/lao/mod.go:127]: yes, will do the update

File: be1-go/channel/mod.go (hunk 7 above)
Comment (MaximeZmt) [reply be1-go/channel/mod.go:72]: I will put it back so :)

**Review by nkcr — APPROVED** [review]

**Pull request merged**